src/snowflake/snowpark/catalog.py

```diff
@@ -0,0 +1,659 @@
+    """
+
+    def __init__(self, session: "snowflake.snowpark.session.Session") -> None:
+        self._session = session
+        self._root = Root(session)
+        self._python_regex_udf = None
+
+    def _parse_database(
+        self,
+        database: Optional[Union[str, Database]],
+        model_obj: Optional[
+            Union[Schema, Table, View, Function, Procedure, UserDefinedFunction]
+        ] = None,
+    ) -> str:
+        if isinstance(
+            model_obj, (Schema, Table, View, Function, Procedure, UserDefinedFunction)
+        ):
+            return model_obj.database_name
+
+        if isinstance(database, str):
+            return database
+        if isinstance(database, Database):
+            return database.name
+        if database is None:
+            return self._session.get_current_database()
+        raise ValueError(
+            f"Unexpected type. Expected str or Database, got '{type(database)}'"
+        )
+
+    def _parse_schema(
+        self,
+        schema: Optional[Union[str, Schema]],
+        model_obj: Optional[
+            Union[Table, View, Function, Procedure, UserDefinedFunction]
+        ] = None,
+    ) -> str:
+        if isinstance(
+            model_obj, (Table, View, Function, Procedure, UserDefinedFunction)
+        ):
+            return model_obj.schema_name
+
+        if isinstance(schema, str):
+            return schema
+        if isinstance(schema, Schema):
+            return schema.name
+        if schema is None:
+            return self._session.get_current_schema()
+        raise ValueError(
+            f"Unexpected type. Expected str or Schema, got '{type(schema)}'"
+        )
+
+    def _parse_function_or_procedure(
+        self,
+        fn: Union[str, Function, Procedure, UserDefinedFunction],
+        arg_types: Optional[List[DataType]],
+    ) -> str:
+        if isinstance(fn, str):
+            if arg_types is None:
+                raise ValueError(
+                    "arg_types must be provided when function/procedure is a string"
+                )
+            arg_types_str = ", ".join(
+                [convert_sp_to_sf_type(arg_type) for arg_type in arg_types]
+            )
+            return f"{fn}({arg_types_str})"
+
+        arg_types_str = ", ".join(arg.datatype for arg in fn.arguments)
+        return f"{fn.name}({arg_types_str})"
+
+    def _initialize_regex_udf(self) -> None:
+        with self._session._lock:
+            if self._python_regex_udf is not None:
+                return
+
+            def python_regex_filter(pattern: str, input: str) -> bool:
+                return bool(re.match(pattern, input))
+
+            self._python_regex_udf = self._session.udf.register(python_regex_filter)
+
+    def _list_objects(
+        self,
+        *,
+        object_name: str,
+        object_class,
+        database: Optional[Union[str, Database]],
+        schema: Optional[Union[str, Schema]],
+        pattern: Optional[str],
```

Comment:
After double-thinking, I suggest we add the `like` parameter as well which is a good snowflake feature.